File: libs/core/pins.cpp
#include "pxt.h"
#include "pins.h"

namespace pins {

    /**
      * Set a pin or connector value to the value
      * @param name pin to write to, eg: DigitalPin.D0
      */
    //% help=pins/digital-write weight=29
    //% blockId=device_set_digital_pin block="digital write |%name=digital_pin|to %value"
    //% value.min=0 value.max=1
    void digitalWrite(int name, int value) {
        ::digitalWrite(name, value);
    }

    /**
     * Set the connector value as analog. Value must be comprised between 0 and 1023.
     * @param name pin name to write to, eg: AnalogPin.A0
     * @param value value to write to the pin between ``0`` and ``1023``. eg:1023,0
     */
    //% help=pins/analog-write weight=24
    //% blockId=device_set_analog_pin block="analog write |%name=analog_pin|to %value"
    //% value.min=0 value.max=1023
    void analogWrite(int name, int value) { 
        ::analogWrite(name, value);
    }

    /**
     * Read the specified pin or connector as either 0 or 1
     * @param name pin to read from, eg: DigitalPin.D0
     */
    //% help=pins/digital-read weight=30
    //% blockId=device_get_digital_pin block="digital read |%name=digital_pin" blockGap=8
    int digitalRead(int name) {
        return ::digitalRead(name);
    }

    /**
     * Read the connector value as analog, that is, as a value comprised between 0 and 1023.
     * @param name pin to write to, eg: AnalogPin.A0
     */
    //% help=pins/analog-read weight=25
    //% blockId=device_get_analog_pin block="analog read |%name=analog_pin" blockGap="8" 
    int analogRead(int name) {
        return ::analogRead(name);
    }

    /**
    * Set the digital pin mode.
    * @param name pin to set the mode on, eg: DigitalPin.D0
    * @param mode one of the pin modes: Input, Output, PullUp, PullDown
    */
    //% help=pins/pin-mode weight=5
    //% blockId=device_set_mode block="set pin |mode %pin=digital_pin|to %mode" blockGap="8" 
    void pinMode(int name, PinMode mode) {
        ::pinMode(name, (int)mode);
    }

    /**
    * Set the analog pin mode.
    * @param name pin to set the mode on, eg: AnalogPin.A0
    * @param mode one of the pin modes: Input, Output, PullUp, PullDown
    */
    //% help=pins/analog-pin-mode weight=4
    //% blockId=device_set_analog_mode block="set analog pin |mode %pin=analog_pin|to %mode"
    void analogPinMode(int name, PinMode mode) {
        ::pinMode(name, (int)mode);
    }

}

File: libs/core/pins.ts

namespace Math {
    /**
     * Re-maps a number from one range to another. That is, a value of ``from low`` would get mapped to ``to low``, a value of ``from high`` to ``to high``, values in-between to values in-between, etc.
     * @param value value to map in ranges
     * @param fromLow the lower bound of the value's current range
     * @param fromHigh the upper bound of the value's current range, eg: 255
     * @param toLow the lower bound of the value's target range
     * @param toHigh the upper bound of the value's target range, eg: 100
     */
    //% help=pins/map weight=23
    //% blockId=math_map block="map %value|from low %fromLow|from high %fromHigh|to low %toLow|to high %toHigh"
    export function map(value: number, fromLow: number, fromHigh: number, toLow: number, toHigh: number): number {
        return ((value - fromLow) * (toHigh - toLow)) / (fromHigh - fromLow) + toLow;
    }
}

namespace pins {

    /**
     * Get the id of a DigitalPin.
     * @param pin the digital pin, eg: DigitalPin.D0
     */
    //% weight=2 help=pins/digital-pin
    //% blockId=digital_pin block="%pin"
    //% shim=TD_ID
    //% pin.defl="0"
    //% useEnumVal=1 blockGap="8"
    export function digitalPin(pin: DigitalPin): number {
        switch (pin) {
            case DigitalPin.D0: return 0;
            case DigitalPin.D1: return 1;
            case DigitalPin.D2: return 2;
            case DigitalPin.D3: return 3;
            case DigitalPin.D4: return 4;
            case DigitalPin.D5: return 5;
        }
        return 0;
    }

    /**
     * Get the id of a AnalogPin.
     * @param pin the analog pin, eg: AnalogPin.D0
     */
    //% weight=1 help=pins/analog-pin
    //% blockId=analog_pin block="%pin"
    //% shim=TD_ID
    //% pin.defl="0"
    //% useEnumVal=1
    export function analogPin(pin: AnalogPin): number {
        switch (pin) {
            case AnalogPin.A0: return 0;
            case AnalogPin.A1: return 1;
            case AnalogPin.A2: return 2;
            case AnalogPin.A3: return 3;
            case AnalogPin.A4: return 4;
            case AnalogPin.A5: return 5;
        }
        return 0;
    }
}

interface Buffer {
    [index: number]: number;
    // rest defined in buffer.cpp
}